src/netaddress.cpp

```diff
@@ -3,63 +3,57 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include <cstdint>
 #include <netaddress.h>
 #include <hash.h>
 #include <util/strencodings.h>
 #include <util/asmap.h>
 #include <tinyformat.h>
 
-static const unsigned char pchIPv4[12] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff };
-static const unsigned char pchOnionCat[] = {0xFD,0x87,0xD8,0x7E,0xEB,0x43};
+#include <algorithm>
+#include <cstdint>
+#include <tuple>
+
+#ifdef WIN32
+#include <winsock.h> // htonl()
+#else
+#include <arpa/inet.h> // htonl()
+#endif
 
-// 0xFD + sha256("bitcoin")[0:5]
-static const unsigned char g_internal_prefix[] = { 0xFD, 0x6B, 0x88, 0xC0, 0x87, 0x24 };
+constexpr size_t CNetAddr::V1_SERIALIZATION_SIZE;
 
 /**
  * Construct an unspecified IPv6 network address (::/128).
  *
  * @note This address is considered invalid by CNetAddr::IsValid()
  */
-CNetAddr::CNetAddr()
-{
-    memset(ip, 0, sizeof(ip));
-}
+CNetAddr::CNetAddr() {}
 
 void CNetAddr::SetIP(const CNetAddr& ipIn)
 {
     m_net = ipIn.m_net;
-    memcpy(ip, ipIn.ip, sizeof(ip));
+    m_addr = ipIn.m_addr;
 }
 
-void CNetAddr::SetLegacyIPv6(const uint8_t ipv6[16])
+void CNetAddr::SetLegacyIPv6(Span<const uint8_t> ipv6)
 {
-    if (memcmp(ipv6, pchIPv4, sizeof(pchIPv4)) == 0) {
+    assert(ipv6.size() == ADDR_IPV6_SIZE);
+
+    if (memcmp(ipv6.begin(), IPV4_IN_IPV6_PREFIX, sizeof(IPV4_IN_IPV6_PREFIX)) == 0) {
```

Comment:
nit, you could reformat this to more "modern" C++ (though I'm not sure if it is less or more readable)
```c++
if(std::equal(std::cbegin(IPV4_IN_IPV6_PREFIX), std::cend(IPV4_IN_IPV6_PREFIX), ipv6.cbegin()))
```

Reply:
Agree with @elichai's suggestion, `std::equal` is nicer.

Given how many of these tests are being done (including the IsRFC... stuff below), perhaps you want to go even further (this is absolutely not a blocker for this PR, however):

```c++
template<typename T, int LEN>
bool HasPrefix(const T1& obj, const uint8_t (&prefix)[LEN])
{
    return obj.size() >= LEN && std::equal(std::cbegin(prefix), std::cend(prefix), std::cbegin(obj));
}

...

if (HasPrefix(ipv6, IPV4_IN_IPV6_PREFIX)) ...
```

If you turn the prefixes into `std::array` it could be
```c++
template<typename T1, typename T2>
bool HasPrefix(const T1& obj, const T2& prefix)
{
    return obj.size() >= prefix.size() && std::equal(std::cbegin(prefix), std::cend(prefix), std::cbegin(obj));
}
```

instead.

Reply:
In C++17 `HasPrefix` will be possible even with C arrays :) (by replacing `.size()` with `std::size(..)`)

Reply:
Done - [diff together with other suggestions](https://github.com/bitcoin/bitcoin/compare/8b7f2c13b67ae84a59f29a7cffe243055a2b93d9..86b2397faedcd9b57649ddb90e98c8cf6a4a6ad3).